core/src/main/scala/org/apache/spark/rpc/netty/Inbox.scala

```diff
@@ -0,0 +1,258 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.rpc.netty
+
+import java.util.concurrent.ConcurrentLinkedQueue
+
+import scala.util.control.NonFatal
+
+import org.apache.spark.{SparkException, Logging}
+import org.apache.spark.rpc.{ThreadSafeRpcEndpoint, RpcAddress, RpcEndpoint}
+
+private[netty] sealed trait InboxMessage
+
+private[netty] case class ContentMessage(
+    senderAddress: RpcAddress,
+    content: Any,
+    needReply: Boolean,
+    context: NettyRpcCallContext) extends InboxMessage
+
+/**
+ * A message type that will be posted to all registered [[RpcEndpoint]]
+ */
+private[netty] sealed trait BroadcastMessage extends InboxMessage
+
+private[netty] case object OnStart extends InboxMessage
+
+private[netty] case object OnStop extends InboxMessage
+
+private[netty] case class Associated(remoteAddress: RpcAddress) extends BroadcastMessage
+
+/**
+ * A broadcast message that indicates
+ */
+private[netty] case class Disassociated(remoteAddress: RpcAddress) extends BroadcastMessage
+
+/**
+ * A broadcast message that indicates a network error
+ */
+private[netty] case class AssociationError(cause: Throwable, remoteAddress: RpcAddress)
+  extends BroadcastMessage
+
+private[netty] abstract class Inbox(
+    val endpointRef: NettyRpcEndpointRef, val endpoint: RpcEndpoint) extends Logging {
+
+  protected val messages = new ConcurrentLinkedQueue[InboxMessage]()
+
+  /**
+   * Process stored messages. Return `true` if the `Inbox` is already stopped, and the caller will
+   * release all resources used by the `Inbox`.
+   */
+  def process(dispatcher: Dispatcher): Boolean
+
+  def post(message: InboxMessage): Unit
+
+  protected def onDrop(message: Any): Unit = {
+    logWarning(s"Drop ${message} because $endpointRef is stopped")
+  }
+
+  def stop(): Unit
+
+  protected def safelyCall(endpoint: RpcEndpoint)(action: => Unit): Unit = {
+    try {
+      action
+    } catch {
+      case NonFatal(e) => {
+        try {
+          endpoint.onError(e)
+        } catch {
+          case NonFatal(e) => logError(s"Ignore error: ${e.getMessage}", e)
+        }
+      }
+    }
+  }
+
+  def isEmpty: Boolean = messages.isEmpty
+
+}
+
+/**
+ * A inbox that stores messages for an [[RpcEndpoint]] and posts messages to it thread-safely.
+ * @param endpointRef
+ * @param endpoint
+ */
+private[netty] class ThreadSafeInbox(
+    endpointRef: NettyRpcEndpointRef,
+    override val endpoint: ThreadSafeRpcEndpoint) extends Inbox(endpointRef, endpoint) {
+
+  private val _endpoint = endpoint.asInstanceOf[ThreadSafeRpcEndpoint]
+
+  // protected by "this"
+  private var stopped = false
+
+  // OnStart should be the first message to process
+  messages.add(OnStart)
```

Comment:
Hmm, I see that you only use `OnStart` and `OnStop` in this inbox. Is there a reason not to use those for both inbox implementations? That would simplify the code a lot.